crates/turborepo-filewatch/src/package_changes_watcher.rs

```diff
@@ -0,0 +1,234 @@
+        Self {
+            _exit_tx: exit_tx,
+            _handle,
+            package_change_events_rx,
+        }
+    }
+
+    pub fn package_changes(&self) -> watch::Receiver<PackageChangeEvent> {
+        self.package_change_events_rx.clone()
+    }
+}
+
+struct Subscriber {
+    file_events_lazy: OptionalWatch<broadcast::Receiver<Result<Event, NotifyError>>>,
+    repo_root: AbsoluteSystemPathBuf,
+    package_change_events_tx: watch::Sender<PackageChangeEvent>,
+}
+
+// This is a workaround because `ignore` doesn't match against a path's
+// ancestors, i.e. if we have `foo/bar/baz` and the .gitignore has `foo/`, it
+// won't match.
+fn ancestors_is_ignored(gitignore: &Gitignore, path: &AnchoredSystemPath) -> bool {
+    path.ancestors().enumerate().any(|(idx, p)| {
+        let is_dir = idx != 0;
+        gitignore.matched(p, is_dir).is_ignore()
+    })
+}
+
+struct RepoState {
+    file_events: broadcast::Receiver<Result<Event, NotifyError>>,
+    pkg_dep_graph: PackageGraph,
+}
+
+impl Subscriber {
+    fn new(
+        repo_root: AbsoluteSystemPathBuf,
+        file_events_lazy: OptionalWatch<broadcast::Receiver<Result<Event, NotifyError>>>,
+        package_change_events_tx: watch::Sender<PackageChangeEvent>,
+    ) -> Self {
+        Subscriber {
+            repo_root,
+            file_events_lazy,
+            package_change_events_tx,
+        }
+    }
+
+    async fn initialize_repo_state(&mut self) -> Option<RepoState> {
+        let Ok(file_events) = self.file_events_lazy.get().await.map(|r| r.resubscribe()) else {
+            // if we get here, it means that file watching has not started, so we should
+            // just report that the package watcher is not available
+            tracing::debug!("file watching shut down, package watcher not available");
+            return None;
+        };
+
+        let Ok(root_package_json) =
+            PackageJson::load(&self.repo_root.join_component("package.json"))
+        else {
+            tracing::debug!("no package.json found, package watcher not available");
+            return None;
+        };
+
+        let Ok(pkg_dep_graph) = PackageGraphBuilder::new(&self.repo_root, root_package_json)
+            .build()
+            .await
+        else {
+            tracing::debug!("package graph not available, package watcher not available");
+            return None;
+        };
+
+        Some(RepoState {
+            file_events,
+            pkg_dep_graph,
+        })
+    }
+
+    async fn watch(mut self, exit_rx: oneshot::Receiver<()>) {
+        let process = async {
+            let Some(mut repo_state) = self.initialize_repo_state().await else {
+                return;
+            };
+
+            self.package_change_events_tx
+                .send(PackageChangeEvent::Rediscover)
+                .ok();
+            // TODO: Pass in global_deps and ignore_patterns
+            let mut change_mapper = ChangeMapper::new(&repo_state.pkg_dep_graph, vec![], vec![]);
```

Comment:
I guess this is where the package discovery piece comes in? We need to invalidate the change mapper on changes to the package graph?

Reply:
Yeah. We're basically building up a [salsa](https://github.com/salsa-rs/salsa) style query system where we have dependent data like the package graph and the change mapper on top of the package graph, that is derived from the files. When the files change, we need to rebuilt the package graph and therefore the change mapper.